File: golang/tests/pkg/unit/mobile_simulate_test.go
package agentbay_test

import (
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// TestMobileSimulateUploadResult tests the MobileSimulateUploadResult struct
func TestMobileSimulateUploadResult(t *testing.T) {
	t.Run("SuccessResult", func(t *testing.T) {
		result := &agentbay.MobileSimulateUploadResult{
			Success:                 true,
			MobileSimulateContextID: "context-123",
		}
		if !result.Success {
			t.Error("Expected Success to be true")
		}
		if result.MobileSimulateContextID != "context-123" {
			t.Errorf("Expected MobileSimulateContextID=context-123, got %s", result.MobileSimulateContextID)
		}
		if result.ErrorMessage != "" {
			t.Errorf("Expected empty ErrorMessage, got %s", result.ErrorMessage)
		}
	})

	t.Run("FailureResult", func(t *testing.T) {
		result := &agentbay.MobileSimulateUploadResult{
			Success:      false,
			ErrorMessage: "Upload failed",
		}
		if result.Success {
			t.Error("Expected Success to be false")
		}
		if result.MobileSimulateContextID != "" {
			t.Errorf("Expected empty MobileSimulateContextID, got %s", result.MobileSimulateContextID)
		}
		if result.ErrorMessage != "Upload failed" {
			t.Errorf("Expected ErrorMessage='Upload failed', got %s", result.ErrorMessage)
		}
	})
}

// TestMobileSimulateServiceInitialization tests the NewMobileSimulateService function
func TestMobileSimulateServiceInitialization(t *testing.T) {
	t.Run("ValidInitialization", func(t *testing.T) {
		agentBayClient := &agentbay.AgentBay{
			Context: &agentbay.ContextService{},
		}
		service, err := agentbay.NewMobileSimulateService(agentBayClient)
		if err != nil {
			t.Fatalf("Expected no error, got %v", err)
		}
		if service == nil {
			t.Fatal("Expected non-nil service")
		}
		// Check default values
		if service.GetSimulateEnable() {
			t.Error("Expected SimulateEnable to be false by default")
		}
		if service.GetSimulateMode() != models.MobileSimulateModePropertiesOnly {
			t.Errorf("Expected default SimulateMode to be PropertiesOnly, got %s", service.GetSimulateMode())
		}
	})

	t.Run("NilAgentBay", func(t *testing.T) {
		service, err := agentbay.NewMobileSimulateService(nil)
		if err == nil {
			t.Error("Expected error for nil AgentBay")
		}
		if service != nil {
			t.Error("Expected nil service")
		}
		if err != nil && err.Error() != "agentBay is required" {
			t.Errorf("Expected 'agentBay is required' error, got %v", err)
		}
	})

	t.Run("NilContext", func(t *testing.T) {
		agentBayClient := &agentbay.AgentBay{
			Context: nil,
		}
		service, err := agentbay.NewMobileSimulateService(agentBayClient)
		if err == nil {
			t.Error("Expected error for nil Context")
		}
		if service != nil {
			t.Error("Expected nil service")
		}
		if err != nil && err.Error() != "agentBay.Context is required" {
			t.Errorf("Expected 'agentBay.Context is required' error, got %v", err)
		}
	})
}

// TestSetAndGetSimulateEnable tests the SetSimulateEnable and GetSimulateEnable methods
func TestSetAndGetSimulateEnable(t *testing.T) {
	agentBayClient := &agentbay.AgentBay{
		Context: &agentbay.ContextService{},
	}
	service, err := agentbay.NewMobileSimulateService(agentBayClient)
	if err != nil {
		t.Fatalf("Failed to create service: %v", err)
	}

	t.Run("DefaultValue", func(t *testing.T) {
		if service.GetSimulateEnable() {
			t.Error("Expected default SimulateEnable to be false")
		}
	})

	t.Run("SetEnableTrue", func(t *testing.T) {
		service.SetSimulateEnable(true)
		if !service.GetSimulateEnable() {
			t.Error("Expected SimulateEnable to be true")
		}
	})

	t.Run("SetEnableFalse", func(t *testing.T) {
		service.SetSimulateEnable(false)
		if service.GetSimulateEnable() {
			t.Error("Expected SimulateEnable to be false")
		}
	})
}

// TestSetAndGetSimulateMode tests the SetSimulateMode and GetSimulateMode methods
func TestSetAndGetSimulateMode(t *testing.T) {
	agentBayClient := &agentbay.AgentBay{
		Context: &agentbay.ContextService{},
	}
	service, err := agentbay.NewMobileSimulateService(agentBayClient)
	if err != nil {
		t.Fatalf("Failed to create service: %v", err)
	}

	t.Run("DefaultMode", func(t *testing.T) {
		if service.GetSimulateMode() != models.MobileSimulateModePropertiesOnly {
			t.Errorf("Expected default SimulateMode to be PropertiesOnly, got %s", service.GetSimulateMode())
		}
	})

	t.Run("SetModeAll", func(t *testing.T) {
		service.SetSimulateMode(models.MobileSimulateModeAll)
		if service.GetSimulateMode() != models.MobileSimulateModeAll {
			t.Errorf("Expected SimulateMode=All, got %s", service.GetSimulateMode())
		}
	})

	t.Run("SetModeSensorsOnly", func(t *testing.T) {
		service.SetSimulateMode(models.MobileSimulateModeSensorsOnly)
		if service.GetSimulateMode() != models.MobileSimulateModeSensorsOnly {
			t.Errorf("Expected SimulateMode=SensorsOnly, got %s", service.GetSimulateMode())
		}
	})

	t.Run("SetModePackagesOnly", func(t *testing.T) {
		service.SetSimulateMode(models.MobileSimulateModePackagesOnly)
		if service.GetSimulateMode() != models.MobileSimulateModePackagesOnly {
			t.Errorf("Expected SimulateMode=PackagesOnly, got %s", service.GetSimulateMode())
		}
	})

	t.Run("SetModeServicesOnly", func(t *testing.T) {
		service.SetSimulateMode(models.MobileSimulateModeServicesOnly)
		if service.GetSimulateMode() != models.MobileSimulateModeServicesOnly {
			t.Errorf("Expected SimulateMode=ServicesOnly, got %s", service.GetSimulateMode())
		}
	})
}

// TestSetAndGetSimulateContextID tests the SetSimulateContextID and GetSimulateContextID methods
func TestSetAndGetSimulateContextID(t *testing.T) {
	agentBayClient := &agentbay.AgentBay{
		Context: &agentbay.ContextService{},
	}
	service, err := agentbay.NewMobileSimulateService(agentBayClient)
	if err != nil {
		t.Fatalf("Failed to create service: %v", err)
	}

	t.Run("DefaultContextID", func(t *testing.T) {
		if service.GetSimulateContextID() != "" {
			t.Errorf("Expected empty default ContextID, got %s", service.GetSimulateContextID())
		}
	})

	t.Run("SetContextID", func(t *testing.T) {
		contextID := "test-context-123"
		service.SetSimulateContextID(contextID)
		if service.GetSimulateContextID() != contextID {
			t.Errorf("Expected ContextID=%s, got %s", contextID, service.GetSimulateContextID())
		}
	})

	t.Run("SetEmptyContextID", func(t *testing.T) {
		service.SetSimulateContextID("")
		if service.GetSimulateContextID() != "" {
			t.Errorf("Expected empty ContextID, got %s", service.GetSimulateContextID())
		}
	})
}

// TestGetSimulateConfig tests the GetSimulateConfig method
func TestGetSimulateConfig(t *testing.T) {
	agentBayClient := &agentbay.AgentBay{
		Context: &agentbay.ContextService{},
	}
	service, err := agentbay.NewMobileSimulateService(agentBayClient)
	if err != nil {
		t.Fatalf("Failed to create service: %v", err)
	}

	t.Run("DefaultConfig", func(t *testing.T) {
		config := service.GetSimulateConfig()
		if config == nil {
			t.Fatal("Expected non-nil config")
		}
		if config.Simulate {
			t.Error("Expected Simulate to be false by default")
		}
		if config.SimulateMode != models.MobileSimulateModePropertiesOnly {
			t.Errorf("Expected default SimulateMode to be PropertiesOnly, got %s", config.SimulateMode)
		}
	})

	t.Run("ConfigWithInternalContext", func(t *testing.T) {
		service.SetSimulateEnable(true)
		service.SetSimulateMode(models.MobileSimulateModeAll)
		service.SetSimulateContextID("internal-ctx-123")

		config := service.GetSimulateConfig()
		if config == nil {
			t.Fatal("Expected non-nil config")
		}
		if !config.Simulate {
			t.Error("Expected Simulate to be true")
		}
		if config.SimulateMode != models.MobileSimulateModeAll {
			t.Errorf("Expected SimulateMode=All, got %s", config.SimulateMode)
		}
		if config.SimulatedContextID != "internal-ctx-123" {
			t.Errorf("Expected SimulatedContextID=internal-ctx-123, got %s", config.SimulatedContextID)
		}
		if config.SimulatePath == "" {
			t.Error("Expected non-empty SimulatePath")
		}
	})

	t.Run("ConfigWithAllModes", func(t *testing.T) {
		modes := []models.MobileSimulateMode{
			models.MobileSimulateModePropertiesOnly,
			models.MobileSimulateModeSensorsOnly,
			models.MobileSimulateModePackagesOnly,
			models.MobileSimulateModeServicesOnly,
			models.MobileSimulateModeAll,
		}

		for _, mode := range modes {
			service.SetSimulateMode(mode)
			config := service.GetSimulateConfig()
			if config.SimulateMode != mode {
				t.Errorf("Expected SimulateMode=%s, got %s", mode, config.SimulateMode)
			}
		}
	})
}

// TestMobileSimulateConfig tests the MobileSimulateConfig model
func TestMobileSimulateConfig(t *testing.T) {
	t.Run("DefaultConfig", func(t *testing.T) {
		config := &models.MobileSimulateConfig{}
		if config.Simulate {
			t.Error("Expected Simulate to be false by default")
		}
		if config.SimulateMode != "" {
			t.Errorf("Expected empty SimulateMode, got %s", config.SimulateMode)
		}
		if config.SimulatedContextID != "" {
			t.Errorf("Expected empty SimulatedContextID, got %s", config.SimulatedContextID)
		}
		if config.SimulatePath != "" {
			t.Errorf("Expected empty SimulatePath, got %s", config.SimulatePath)
		}
	})

	t.Run("ConfigWithAllFields", func(t *testing.T) {
		config := &models.MobileSimulateConfig{
			Simulate:           true,
			SimulatePath:       "/data/agentbay_mobile_info",
			SimulateMode:       models.MobileSimulateModeAll,
			SimulatedContextID: "ctx-123",
		}
		if !config.Simulate {
			t.Error("Expected Simulate to be true")
		}
		if config.SimulatePath != "/data/agentbay_mobile_info" {
			t.Errorf("Expected SimulatePath=/data/agentbay_mobile_info, got %s", config.SimulatePath)
		}
		if config.SimulateMode != models.MobileSimulateModeAll {
			t.Errorf("Expected SimulateMode=All, got %s", config.SimulateMode)
		}
		if config.SimulatedContextID != "ctx-123" {
			t.Errorf("Expected SimulatedContextID=ctx-123, got %s", config.SimulatedContextID)
		}
	})

	t.Run("AllSimulateModes", func(t *testing.T) {
		modes := []models.MobileSimulateMode{
			models.MobileSimulateModePropertiesOnly,
			models.MobileSimulateModeSensorsOnly,
			models.MobileSimulateModePackagesOnly,
			models.MobileSimulateModeServicesOnly,
			models.MobileSimulateModeAll,
		}

		expectedValues := []string{
			"PropertiesOnly",
			"SensorsOnly",
			"PackagesOnly",
			"ServicesOnly",
			"All",
		}

		for i, mode := range modes {
			if string(mode) != expectedValues[i] {
				t.Errorf("Expected mode %s, got %s", expectedValues[i], string(mode))
			}
		}
	})
}

// TestMobileExtraConfigWithSimulate tests MobileExtraConfig with SimulateConfig
func TestMobileExtraConfigWithSimulate(t *testing.T) {
	t.Run("WithSimulateConfig", func(t *testing.T) {
		simulateConfig := &models.MobileSimulateConfig{
			Simulate:           true,
			SimulateMode:       models.MobileSimulateModeAll,
			SimulatedContextID: "ctx-456",
			SimulatePath:       "/data/agentbay_mobile_info",
		}

		mobileConfig := &models.MobileExtraConfig{
			LockResolution: true,
			SimulateConfig: simulateConfig,
		}

		if mobileConfig.SimulateConfig == nil {
			t.Fatal("Expected non-nil SimulateConfig")
		}
		if !mobileConfig.SimulateConfig.Simulate {
			t.Error("Expected Simulate to be true")
		}
		if mobileConfig.SimulateConfig.SimulateMode != models.MobileSimulateModeAll {
			t.Errorf("Expected SimulateMode=All, got %s", mobileConfig.SimulateConfig.SimulateMode)
		}
		if mobileConfig.SimulateConfig.SimulatedContextID != "ctx-456" {
			t.Errorf("Expected SimulatedContextID=ctx-456, got %s", mobileConfig.SimulateConfig.SimulatedContextID)
		}
	})

	t.Run("WithoutSimulateConfig", func(t *testing.T) {
		mobileConfig := &models.MobileExtraConfig{
			LockResolution: true,
		}

		if mobileConfig.SimulateConfig != nil {
			t.Error("Expected nil SimulateConfig")
		}
	})
}

// TestMobileSimulateConstants tests the mobile simulate constants
func TestMobileSimulateConstants(t *testing.T) {
	t.Run("PathConstants", func(t *testing.T) {
		// Check that constants are defined (values from config.go)
		if agentbay.MobileInfoSubPath != "/agentbay_mobile_info/" {
			t.Errorf("Expected MobileInfoSubPath=/agentbay_mobile_info/, got %s", agentbay.MobileInfoSubPath)
		}
		if agentbay.MobileInfoFileName != "dev_info.json" {
			t.Errorf("Expected MobileInfoFileName=dev_info.json, got %s", agentbay.MobileInfoFileName)
		}
		if agentbay.MobileInfoDefaultPath != "/data/agentbay_mobile_info" {
			t.Errorf("Expected MobileInfoDefaultPath=/data/agentbay_mobile_info, got %s", agentbay.MobileInfoDefaultPath)
		}
	})
}

// TestUploadMobileInfoValidation tests the UploadMobileInfo validation
func TestUploadMobileInfoValidation(t *testing.T) {
	agentBayClient := &agentbay.AgentBay{
		Context: &agentbay.ContextService{},
	}
	service, err := agentbay.NewMobileSimulateService(agentBayClient)
	if err != nil {
		t.Fatalf("Failed to create service: %v", err)
	}

	t.Run("EmptyContent", func(t *testing.T) {
		result := service.UploadMobileInfo("", nil)
		if result.Success {
			t.Error("Expected upload to fail with empty content")
		}
		if result.ErrorMessage == "" {
			t.Error("Expected error message")
		}
		if result.ErrorMessage != "mobileDevInfoContent is required" {
			t.Errorf("Expected 'mobileDevInfoContent is required', got %s", result.ErrorMessage)
		}
	})

	t.Run("InvalidJSON", func(t *testing.T) {
		result := service.UploadMobileInfo("not a json", nil)
		if result.Success {
			t.Error("Expected upload to fail with invalid JSON")
		}
		if result.ErrorMessage == "" {
			t.Error("Expected error message")
		}
		if !contains(result.ErrorMessage, "not a valid JSON string") {
			t.Errorf("Expected error message to contain 'not a valid JSON string', got %s", result.ErrorMessage)
		}
	})

	t.Run("MissingContextIDInContextSync", func(t *testing.T) {
		contextSync := &agentbay.ContextSync{
			ContextID: "",
			Path:      "/data",
		}
		result := service.UploadMobileInfo(`{"test":"data"}`, contextSync)
		if result.Success {
			t.Error("Expected upload to fail with empty ContextID")
		}
		if result.ErrorMessage != "contextSync.ContextID is required" {
			t.Errorf("Expected 'contextSync.ContextID is required', got %s", result.ErrorMessage)
		}
	})
}

// TestHasMobileInfoValidation tests the HasMobileInfo validation
func TestHasMobileInfoValidation(t *testing.T) {
	agentBayClient := &agentbay.AgentBay{
		Context: &agentbay.ContextService{},
	}
	service, err := agentbay.NewMobileSimulateService(agentBayClient)
	if err != nil {
		t.Fatalf("Failed to create service: %v", err)
	}

	t.Run("NilContextSync", func(t *testing.T) {
		_, err := service.HasMobileInfo(nil)
		if err == nil {
			t.Error("Expected error for nil ContextSync")
		}
		if err.Error() != "contextSync is required" {
			t.Errorf("Expected 'contextSync is required', got %v", err)
		}
	})

	t.Run("EmptyContextID", func(t *testing.T) {
		contextSync := &agentbay.ContextSync{
			ContextID: "",
			Path:      "/data",
		}
		_, err := service.HasMobileInfo(contextSync)
		if err == nil {
			t.Error("Expected error for empty ContextID")
		}
		if err.Error() != "contextSync.ContextID is required" {
			t.Errorf("Expected 'contextSync.ContextID is required', got %v", err)
		}
	})

	t.Run("EmptyPath", func(t *testing.T) {
		contextSync := &agentbay.ContextSync{
			ContextID: "ctx-123",
			Path:      "",
		}
		_, err := service.HasMobileInfo(contextSync)
		if err == nil {
			t.Error("Expected error for empty Path")
		}
		if err.Error() != "contextSync.Path is required" {
			t.Errorf("Expected 'contextSync.Path is required', got %v", err)
		}
	})
}

// Helper function to check if a string contains a substring
func contains(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || len(substr) == 0 ||
		(len(s) > 0 && len(substr) > 0 && findSubstring(s, substr)))
}

func findSubstring(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

File: typescript/tests/unit/mobile-simulate.test.ts
/**
 * Unit tests for MobileSimulateService module.
 * 
 * This test suite covers all functionality of the MobileSimulateService class,
 * including initialization, configuration management, mobile info operations,
 * and context management.
 */

import { MobileSimulateService, type MobileSimulateUploadResult } from '../../src/mobile-simulate';
import { ContextSync } from '../../src/context-sync';
import { Context, ContextService } from '../../src/context';
import { MobileSimulateMode } from '../../src/types/extra-configs';

// Mock AgentBay
interface MockAgentBay {
  context: ContextService;
}

// Mock ContextService
class MockContextService {
  get = jest.fn();
  listFiles = jest.fn();
  getFileUploadUrl = jest.fn();
}

// Mock node-fetch
jest.mock('node-fetch', () => jest.fn());

import fetch from 'node-fetch';

describe('MobileSimulateService', () => {
  let mockAgentBay: MockAgentBay;
  let mockContextService: MockContextService;
  let mobileSimulateService: MobileSimulateService;

  beforeEach(() => {
    mockContextService = new MockContextService();
    mockAgentBay = {
      context: mockContextService as any
    };
    mobileSimulateService = new MobileSimulateService(mockAgentBay as any);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('constructor', () => {
    test('should throw error if agentBay is not provided', () => {
      expect(() => new MobileSimulateService(null as any)).toThrow('agentBay is required');
    });

    test('should throw error if agentBay.context is not provided', () => {
      expect(() => new MobileSimulateService({} as any)).toThrow('agentBay.context is required');
    });

    test('should create instance successfully with valid agentBay', () => {
      expect(mobileSimulateService).toBeInstanceOf(MobileSimulateService);
    });

    test('should initialize with default values', () => {
      expect(mobileSimulateService.getSimulateEnable()).toBe(false);
      expect(mobileSimulateService.getSimulateMode()).toBe(MobileSimulateMode.PropertiesOnly);
      expect(mobileSimulateService.getSimulateContextId()).toBeUndefined();
    });
  });

  describe('setSimulateEnable and getSimulateEnable', () => {
    test('should set and get simulate enable flag', () => {
      mobileSimulateService.setSimulateEnable(true);
      expect(mobileSimulateService.getSimulateEnable()).toBe(true);

      mobileSimulateService.setSimulateEnable(false);
      expect(mobileSimulateService.getSimulateEnable()).toBe(false);
    });

    test('should default to false', () => {
      expect(mobileSimulateService.getSimulateEnable()).toBe(false);
    });
  });

  describe('setSimulateMode and getSimulateMode', () => {
    test('should set and get simulate mode', () => {
      mobileSimulateService.setSimulateMode(MobileSimulateMode.All);
      expect(mobileSimulateService.getSimulateMode()).toBe(MobileSimulateMode.All);

      mobileSimulateService.setSimulateMode(MobileSimulateMode.SensorsOnly);
      expect(mobileSimulateService.getSimulateMode()).toBe(MobileSimulateMode.SensorsOnly);
    });

    test('should default to PropertiesOnly', () => {
      expect(mobileSimulateService.getSimulateMode()).toBe(MobileSimulateMode.PropertiesOnly);
    });

    test('should support all simulate modes', () => {
      const modes = [
        MobileSimulateMode.PropertiesOnly,
        MobileSimulateMode.SensorsOnly,
        MobileSimulateMode.PackagesOnly,
        MobileSimulateMode.ServicesOnly,
        MobileSimulateMode.All
      ];

      modes.forEach(mode => {
        mobileSimulateService.setSimulateMode(mode);
        expect(mobileSimulateService.getSimulateMode()).toBe(mode);
      });
    });
  });

  describe('setSimulateContextId and getSimulateContextId', () => {
    test('should set and get simulate context id', () => {
      const contextId = 'test-context-123';
      mobileSimulateService.setSimulateContextId(contextId);
      expect(mobileSimulateService.getSimulateContextId()).toBe(contextId);
    });

    test('should default to undefined', () => {
      expect(mobileSimulateService.getSimulateContextId()).toBeUndefined();
    });

    test('should handle empty string', () => {
      mobileSimulateService.setSimulateContextId('');
      expect(mobileSimulateService.getSimulateContextId()).toBe('');
    });
  });

  describe('getSimulateConfig', () => {
    test('should return default config', () => {
      const config = mobileSimulateService.getSimulateConfig();
      
      expect(config.simulate).toBe(false);
      expect(config.simulateMode).toBe(MobileSimulateMode.PropertiesOnly);
      expect(config.simulatedContextId).toBeUndefined();
      expect(config.simulatePath).toBeUndefined();
    });

    test('should return config with all settings', () => {
      mobileSimulateService.setSimulateEnable(true);
      mobileSimulateService.setSimulateMode(MobileSimulateMode.All);
      mobileSimulateService.setSimulateContextId('test-ctx-456');

      const config = mobileSimulateService.getSimulateConfig();
      
      expect(config.simulate).toBe(true);
      expect(config.simulateMode).toBe(MobileSimulateMode.All);
      expect(config.simulatedContextId).toBe('test-ctx-456');
    });

    test('should update config when settings change', () => {
      mobileSimulateService.setSimulateEnable(true);
      let config = mobileSimulateService.getSimulateConfig();
      expect(config.simulate).toBe(true);

      mobileSimulateService.setSimulateEnable(false);
      config = mobileSimulateService.getSimulateConfig();
      expect(config.simulate).toBe(false);
    });
  });

  describe('hasMobileInfo', () => {
    test('should throw error if contextSync is not provided', async () => {
      await expect(mobileSimulateService.hasMobileInfo(null as any)).rejects.toThrow('contextSync is required');
    });

    test('should throw error if contextSync.contextId is not provided', async () => {
      const contextSync = new ContextSync('', '/test/path');
      await expect(mobileSimulateService.hasMobileInfo(contextSync)).rejects.toThrow('contextSync.contextId is required');
    });

    test('should throw error if contextSync.path is not provided', async () => {
      const contextSync = new ContextSync('test-context-id', '');
      await expect(mobileSimulateService.hasMobileInfo(contextSync)).rejects.toThrow('contextSync.path is required');
    });

    test('should return true if mobile dev info file exists', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.listFiles.mockResolvedValue({
        success: true,
        entries: [
          { fileName: 'dev_info.json' }
        ]
      });

      const result = await mobileSimulateService.hasMobileInfo(contextSync);

      expect(result).toBe(true);
      expect(mockContextService.listFiles).toHaveBeenCalledWith(
        'test-context-id',
        '/test/path/agentbay_mobile_info/',
        1,
        50
      );
    });

    test('should return false if mobile dev info file does not exist', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.listFiles.mockResolvedValue({
        success: true,
        entries: [
          { fileName: 'other_file.json' }
        ]
      });

      const result = await mobileSimulateService.hasMobileInfo(contextSync);

      expect(result).toBe(false);
    });

    test('should return false if listFiles fails', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.listFiles.mockResolvedValue({
        success: false,
        errorMessage: 'Failed to list files'
      });

      const result = await mobileSimulateService.hasMobileInfo(contextSync);

      expect(result).toBe(false);
    });

    test('should return false if entries is empty', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.listFiles.mockResolvedValue({
        success: true,
        entries: []
      });

      const result = await mobileSimulateService.hasMobileInfo(contextSync);

      expect(result).toBe(false);
    });
  });

  describe('uploadMobileInfo', () => {
    const validMobileInfo = JSON.stringify({ model: 'SM-A505F', brand: 'Samsung' });

    test('should throw error if mobileDevInfoContent is not provided', async () => {
      await expect(mobileSimulateService.uploadMobileInfo('')).rejects.toThrow('mobileDevInfoContent is required');
    });

    test('should throw error if mobileDevInfoContent is not valid JSON', async () => {
      await expect(mobileSimulateService.uploadMobileInfo('invalid json')).rejects.toThrow('mobileDevInfoContent is not a valid JSON string');
    });

    test('should throw error if mobileDevInfoContent is malformed JSON', async () => {
      await expect(mobileSimulateService.uploadMobileInfo('{"incomplete":')).rejects.toThrow('mobileDevInfoContent is not a valid JSON string');
    });

    test('should create new context if not provided', async () => {
      mockContextService.get.mockResolvedValue({
        success: true,
        context: new Context('new-context-id', 'mobile_sim_test_123')
      });
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: true,
        status: 200
      } as any);

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo);

      expect(result.success).toBe(true);
      expect(result.mobileSimulateContextId).toBe('new-context-id');
      expect(mockContextService.get).toHaveBeenCalled();
      const callArgs = mockContextService.get.mock.calls[0];
      expect(callArgs[0]).toContain('mobile_sim_');
      expect(callArgs[1]).toBe(true);
    });

    test('should use provided context sync', async () => {
      const contextSync = new ContextSync('existing-context-id', '/test/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: true,
        status: 200
      } as any);

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync);

      expect(result.success).toBe(true);
      expect(result.mobileSimulateContextId).toBe('existing-context-id');
      expect(mockContextService.get).not.toHaveBeenCalled();
    });

    test('should throw error if contextSync is provided without contextId', async () => {
      const contextSync = new ContextSync('', '/test/path');

      await expect(mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync)).rejects.toThrow('contextSync.contextId is required');
    });

    test('should return error if failed to create context', async () => {
      mockContextService.get.mockResolvedValue({
        success: false,
        errorMessage: 'Failed to create context'
      });

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo);

      expect(result.success).toBe(false);
      expect(result.errorMessage).toBe('Failed to create context for simulate');
    });

    test('should return error if failed to get upload URL', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: false,
        errorMessage: 'Failed to get upload URL'
      });

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync);

      expect(result.success).toBe(false);
      expect(result.errorMessage).toBe('Failed to get upload URL');
    });

    test('should return error if upload fails', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockRejectedValue(new Error('Network error'));

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync);

      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain('Network error');
    });

    test('should return error if upload response is not ok', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: false,
        status: 500,
        statusText: 'Internal Server Error'
      } as any);

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync);

      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain('Upload failed with status 500');
    });

    test('should upload successfully with valid data', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: true,
        status: 200
      } as any);

      const result = await mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync);

      expect(result.success).toBe(true);
      expect(result.mobileSimulateContextId).toBe('test-context-id');
      const fetchCall = (fetch as jest.MockedFunction<typeof fetch>).mock.calls[0];
      expect(fetchCall[0]).toBe('https://test-upload-url.com');
      expect(fetchCall[1]).toHaveProperty('method', 'PUT');
      expect(fetchCall[1]).toHaveProperty('body');
      if (fetchCall[1]) {
        expect(fetchCall[1].body).toBeInstanceOf(Uint8Array);
      }
    });

    test('should upload with correct file path', async () => {
      const contextSync = new ContextSync('test-context-id', '/test/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: true,
        status: 200
      } as any);

      await mobileSimulateService.uploadMobileInfo(validMobileInfo, contextSync);

      expect(mockContextService.getFileUploadUrl).toHaveBeenCalledWith(
        'test-context-id',
        '/test/path/agentbay_mobile_info/dev_info.json'
      );
    });
  });

  describe('integration scenarios', () => {
    test('should handle complete workflow: configure -> upload -> get config', async () => {
      // Configure
      mobileSimulateService.setSimulateEnable(true);
      mobileSimulateService.setSimulateMode(MobileSimulateMode.All);

      // Upload
      const contextSync = new ContextSync('workflow-context-id', '/workflow/path');
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: true,
        status: 200
      } as any);

      const uploadResult = await mobileSimulateService.uploadMobileInfo(
        JSON.stringify({ model: 'TestDevice' }),
        contextSync
      );

      expect(uploadResult.success).toBe(true);

      // Get config
      const config = mobileSimulateService.getSimulateConfig();
      expect(config.simulate).toBe(true);
      expect(config.simulateMode).toBe(MobileSimulateMode.All);
      expect(config.simulatedContextId).toBeUndefined(); // External context
    });

    test('should handle workflow with internal context', async () => {
      mobileSimulateService.setSimulateEnable(true);
      mobileSimulateService.setSimulateMode(MobileSimulateMode.PropertiesOnly);

      mockContextService.get.mockResolvedValue({
        success: true,
        context: new Context('internal-context-id', 'mobile_sim_internal')
      });
      mockContextService.getFileUploadUrl.mockResolvedValue({
        success: true,
        url: 'https://test-upload-url.com'
      });
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValue({
        ok: true,
        status: 200
      } as any);

      const uploadResult = await mobileSimulateService.uploadMobileInfo(
        JSON.stringify({ model: 'InternalDevice' })
      );

      expect(uploadResult.success).toBe(true);
      expect(uploadResult.mobileSimulateContextId).toBe('internal-context-id');

      const config = mobileSimulateService.getSimulateConfig();
      expect(config.simulatedContextId).toBe('internal-context-id'); // Internal context
    });
  });
});
